bootstrap/004_labels.sql

```diff
@@ -114,38 +113,122 @@ SELECT *,$$;
     return true;
 END;
 
-CREATE OR REPLACE PROCEDURE ADMIN.CREATE_LABEL(name text, grp text, rank number, condition text, is_dynamic boolean)
-    returns text
-    language python
-    runtime_version = "3.10"
-    handler = 'create_label'
-    packages = ('snowflake-snowpark-python', 'pydantic==1.*', 'snowflake-telemetry-python')
-    imports = ('{{stage}}/python/crud.zip')
-    EXECUTE AS OWNER
+CREATE OR REPLACE FUNCTION VALIDATION.IS_NULL(name text)
+    RETURNS text
+    COMMENT='Validates that a field in an object is null'
+AS
+$$
+    'TO_CHAR(f:' || name || ') is null'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.NOT_NULL(name text)
+    RETURNS text
+    COMMENT='Validates that a field in an object is not null'
+AS
+$$
+    'TO_CHAR(f:' || name || ') is not null'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.LABEL_EXISTS()
+    RETURNS text
+AS
+$$
+    '(select count(*) > 0 from INTERNAL.LABELS where group_name is null and name = f:name)'
+$$;
+
+CREATE OR REPLACE FUNCTION VALIDATION.LABEL_ABSENT()
+    RETURNS text
 AS
 $$
-from crud import create_entity
-from datetime import datetime
-from uuid import uuid4
-def create_label(session, name, grp, rank, condition, is_dynamic):
-    return create_entity(session, 'LABEL', {'name': name, 'group_name': grp, 'group_rank': rank, 'condition': condition, 'is_dynamic': is_dynamic, 'label_created_at': datetime.now(), 'label_modified_at': datetime.now(), 'label_id': str(uuid4())})
+    'NOT(' || VALIDATION.LABEL_EXISTS() || ')'
+$$;
+
+
+CREATE OR REPLACE PROCEDURE VALIDATION.VALID_LABEL_CONDITION(input varchar)
+    RETURNS BOOLEAN
+    LANGUAGE SQL
+AS
+BEGIN
+    let c varchar := (select parse_json(:input):condition);
+    let stmt varchar := (select 'select case when ' || :c || ' then true else false end from reporting.enriched_query_history limit 1');
+    execute immediate stmt;
+    return true;
```

Comment:
should this return the output of execute immediate?

Reply:
i think we rarely hit the else branch (often raise an exception), but yes that would be correct to do.

Reply:
oh, no, that's not right. We are just testing the compilation here. The actual row we get from query history may evaluate to true or false..